### PR TITLE
feat: added article on Branded Types

### LINKED ISSUE #133: Article: Branded Types
## Overview

The topic of 'branded types' keeps coming up once in a while. There aren't too many resources on it right now - and no article is as deep on them as I want. Time to write it myself!

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #133
- [x] That issue was marked as [accepting prs](https://github.com/LearningTypeScript/site/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/LearningTypeScript/site/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

~Got through about half of it so far.~ ✅ This is complete.

Once typescript-eslint's `no-floating-promises` augmentations are shipped, I'll clean this up and put it in review. ~Tentatively early-mid-June or so.~ early-August it is!

The only blocker to publish is that I'd like `ts-brand` to be published with its latest `make` additions: https://github.com/kourge/ts-brand/issues/11.

### CODE REVIEW

**Review by chriskrycho — COMMENTED** [review]
One alternative you don’t mention is biting the bullet in increased cost and providing a wrapper type, e.g.:

```ts
class Positive {
  #value: number;
  
  private constructor(value: number) {
    this.#value = value;
  }

  static of(value: number) -> Positive | DescriptiveError {
    return value > 0 ? new this(value) : new DescriptiveError(value);
  }
}
```

This obviously has downsides, but in many real-world cases they pay for themselves by way of the fact that you do actually get safety at all points, including a type safe error path.

Bonus: It might or might not make sense to link to Parse, Don’t Validate for additional reading?

Overall, I think this is a great write-up to point people to, regardless of the above potential addition!

**Comment by JoshuaKGoldberg** [comment]
@chriskrycho good point on the wrapper! Added a note with a link to your comment, thanks.

For Don't Parse, Validate, that's https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate, right? If so: great article, but I'm hesitant to include a Haskell-oriented one in this post. I've found that the jump between languages turns away devs who aren't confident enough to make that jump. If there's a popular equivalent for JS/TS-land I'd love that!

File: src/content/articles/branded-types.mdx
Comment (JoshuaKGoldberg) [inline src/content/articles/branded-types.mdx:1]: From @nicojs (https://x.com/_nicojs/status/1818339638349594685):

I wouldn't call type guards 'type assertion functions'. A type assertion function would be:

function positive(n: number): asserts n is Positive {
  if(n<0) { 
     throw new TypeError(`${n} is not a positive number`);
  }
}

**Comment by chriskrycho** [comment]
@JoshuaKGoldberg yeah, that’s the one, and I agree that dumping folks into Haskell can feel like a *lot*. 😂 I don’t know of a good TS-focused write-up, alas (most of the ones I have found poking around are not great in various ways). It probably deserves its own entry here, honestly!

**Pull request merged**